### PR TITLE
fix(profiling): vendor protobuf

### PR DESCRIPTION
**Pull request by emmettbutler**
This change fixes issues like https://github.com/DataDog/dd-trace-py/pull/3766 in which incompatibility can arise between the version of protobuf that ddtrace brings with it and the version that the customer's app uses. In practice these issues occur under lib-injection, because in such scenarios ddtrace's requirements are finalized before the app's requirements are known. The reason this change fixes such issues is that it avoids using the `google.protobuf` module that the application may care about, instead using its own `ddtrace.vendor.protobuf.google.protobuf` module.

The fix is to vendor the most recent release of protobuf, and use the vendored version for all Python runtime versions that support it. Experimentally, this set of versions seems to be 3.8+. The behavior of the protobuf library under 3.7 is unchanged in this pull request. Vendoring protobuf for 3.7 would require adding another copy of the protobuf code to the repo.

The code in `vendor/protobuf` was taken from an unzipped wheel of 5.26.0 downloaded from PyPI. I chose this method instead of building from source in setup.py because protobuf has a complex build process managed by Bazel, and I didn't want to add Bazel as a dependency of ddtrace's build process.

There are changes to the riotfile to stop installing protobuf in test environments that will use the vendored version. There is also a new `pprof` file added that imports the vendored code. This file was created by copypasting the existing `pprof` file with the highest number and changing the imports.

Passing existing functional tests is sufficient because there's no new functionality added here.

Files you can ignore in code review:
* `ddtrace/vendor`
* `*_pb2*`
* Riot lockfiles

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [x] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [ ] Title is accurate
- [ ] All changes are related to the pull request's stated goal
- [ ] Description motivates each change
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [ ] Testing strategy adequately addresses listed risks
- [ ] Change is maintainable (easy to change, telemetry, documentation)
- [ ] Release note makes sense to a user of the library
- [ ] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [ ] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -145,6 +145,7 @@ exclude-modules = '''
   | ddtrace.profiling.exporter.pprof_312_pb2
```
Comment (Kyle-Verhoog) [inline pyproject.toml:145]: (non-blocking) we can get rid of these files after this change, right?

Comment (Kyle-Verhoog) [reply pyproject.toml:145]: oh, nvm we still need them for Python < 3.8.

File: pyproject.toml [hunk 2]
```diff
@@ -34,7 +34,7 @@ dependencies = [
     "envier~=0.5",
```
Comment (Kyle-Verhoog) [inline pyproject.toml:34]: Can't comment on the diff on the ddsketch line, but I believe ddsketch has protobuf dependency as well that we'll need to deal with: https://github.com/DataDog/sketches-py/blob/c0b33bcf2dadbd0b16ee44cecfcceae86e93fd6c/setup.py#L23-L24. ddsketch is only used when tracer computed stats is enabled.

**Review by Kyle-Verhoog — COMMENTED** [review]
Awesome, this change ends ends up being surprisingly minimal in terms of change to ddtrace!

I have a concern about ddsketch having a dependency on protobuf which will still be pulled in. I think we could resolve it by storing the ddsketch protobuf in ddtrace and relaxing the protobuf dependency of the ddsketch package.

**Review by sanchda — APPROVED** [review]

File: pyproject.toml (hunk 2 above)
Comment (emmettbutler) [reply pyproject.toml:34]: https://github.com/DataDog/sketches-py/pull/62

File: ddtrace/vendor/protobuf/google/protobuf/internal/api_implementation.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    # Work around a known issue in the classic bootstrap .par import hook.
+    if not mod:
+      raise ImportError(mod_name + ' import succeeded but was None')
+    return True
+  except ImportError:
+    return False
+
+
+if _implementation_type is None:
+  if _CanImport('google._upb._message'):
```
Comment (brettlangdon) [inline ddtrace/vendor/protobuf/google/protobuf/internal/api_implementation.py:53]: any checks or imports like this need to be updatd.

**Review by brettlangdon — CHANGES_REQUESTED** [review]

**Comment by emmettbutler** [comment]
New plan, after discovering that vendoring protobuf requires a lot more work than previously thought:
@sanchda will update `lib-injection/dl_wheels.py` such that it avoids loading protobuf whenever possible (under lib-injection and an alternate profiler exporter implementation is available).

**Pull request closed without merging**